### ISSUE TITLE
Actions workflow for profiling unit test runs

### ISSUE THREAD

**Issue by sadielbartholomew** [NCAS-CMS/cf-python#127]
It would be useful to have a CI workflow that runs cProfile on the unit test run (or perhaps some custom script that is more useful to profile) for the jobs with their differing matrix variables (currently OS & Python version), that we can run ad-hoc.

It would allow us to pinpoint identify performance differences across platforms and Python versions. In particular I raise this since I have noticed the MacOS jobs all consistently run significantly slower than the Ubuntu jobs over the current test suite. With profiling stats we could investigate that with detailed outputs and without having to rummage through the job logs for timings. Otherwise it is hard to check up on potential Mac-specific bottlenecks, not having a Mac at hand.

Ideally we can have a workflow that grabs the profiling stats for unit test runs with each matrix variable and then runs a final step with some logic to determine any major differences in outputs, to inform us of any significant  platform or version-specific differences. Or runs a vis tool like snakeviz or gprof2dot to eyeball for noticeable differences.

**Comment by davidhassell** [comment]
Sounds like a good idea. I presume that by "ad hoc" you mean we would choose when to run this action, rather than it running automatically every time the (other) actions are triggered?

**Comment by sadielbartholomew** [comment]
Yes, just when we think it would be useful to profile. Unless you prefer otherwise? Possibly semi-regularly is a better approach. Certainly not every single commit (even it is just essentially wrapping the unit test run in cProfile, I think there is already enough noise in the workflow), but every now and again to check nothing is wildly slow or platform-dependent.

**Comment by davidhassell** [comment]
Ad-hoc or semi-regular sounds fine to me.
